src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+
+        task.stack[STACK_SIZE - 13] = 0; // edi
+        task.stack[STACK_SIZE - 12] = 0; // esi
+        // End of the stack
+        task.stack[STACK_SIZE - 11] = @ptrToInt(&task.stack[STACK_SIZE - 1]); // ebp
+        task.stack[STACK_SIZE - 10] = 0; // esp (temp) this won't be popped by popa bc intel is dump XD
+
+        task.stack[STACK_SIZE - 9] = 0; // ebx
+        task.stack[STACK_SIZE - 8] = 0; // edx
+        task.stack[STACK_SIZE - 7] = 0; // ecx
+        task.stack[STACK_SIZE - 6] = 0; // eax
+
+        task.stack[STACK_SIZE - 5] = 0; // int_num
+        task.stack[STACK_SIZE - 4] = 0; // error_code
+
+        task.stack[STACK_SIZE - 3] = @ptrToInt(entry_point); // eip
+        task.stack[STACK_SIZE - 2] = gdt.KERNEL_CODE_OFFSET; // cs
+        task.stack[STACK_SIZE - 1] = 0x202; // eflags
+
+        // Point to top of the created stack above
+        task.esp = @ptrToInt(&task.stack[STACK_SIZE - 18]);
+
+        return task;
+    }
+
+    ///
+    /// Destroy the task. This will release the allocated PID and free the stack and self.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///
+    pub fn destroy(self: *Self) void {
+        freePid(self.pid);
+        alloc.free(self.stack);
+        alloc.destroy(self);
+    }
+
+    ///
+    /// Save a new stack pointer for this task.
+    ///
+    /// Arguments:
+    ///     IN/OUT self: *Self - The pointer to self.
+    ///     IN esp: u32        - The new stack pointer to be saved.
+    ///
+    pub fn saveEsp(self: *Self, esp: u32) void {
+        self.esp = esp;
+    }
+};
+
+///
+/// Allocate a process identifier. If out of PIDs, then will panic. Is this occurs, will need to
+/// increase the bitmap.
+///
+/// Return: u32
+///     A new PID.
+///
+fn allocatePid() u32 {
+    return all_pids.setFirstFree() orelse panic(@errorReturnTrace(), "Out of PIDs", .{});
+}
+
+///
+/// Free an allocated PID. One must be allocated to be freed. If one wasn't allocated will panic.
+///
+/// Arguments:
+///     IN pid: u32 - The PID to free.
+///
+fn freePid(pid: u32) void {
+    all_pids.clearEntry(pid) catch panic(@errorReturnTrace(), "PID {} not allocated", .{pid});
+}
+
+///
+/// Round robin. This will pick the next task to be run from the queue. It will add the current
+/// task to the end of the queue and pop the next task from the front as set this as the current
+/// task. Interrupts are assumed disabled.
+///
+fn pickNextTask() void {
+    // Get the next task
+    if (tasks.pop()) |new_task_node| {
+        var next_task = new_task_node.data;
+        tasks.destroyNode(new_task_node, alloc);
+
+        // Now that we have the new task and deleted its node, there "will" be space for another
+        // so the allocation shouldn't fail as interrupts are disabled.
+        // So if there is a failure, then wow :P This will be the case for multiprocessor, but
+        // for a later time
```

Comment:
I think `catch unreachable` is justified in this case, but that can come down to taste :)

Reply:
I guess so yep, I am just reluctant in using unreachable as this is removed in releasefast builds and would lead to undefined behaviour

Reply:
True! I'm OK with whatever you choose.

Reply:
I have changed this so no allocations are needed